dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -260,71 +209,115 @@ private void simplifyAssociatives(Class<? extends Definition> cls, BiFunction<Re
+            if (!relevantRelations.containsAll(c.getConstrainedRelations())) {
+                removeConstraint(c);
+            }
+        }
+
+        for (Relation rel : Set.copyOf(getRelations())) {
+            if (!relevantRelations.contains(rel)) {
+                deleteRelation(rel);
+            }
+        }
+}
+
+    private final static class DependencyCollector implements Constraint.Visitor<Void> {
+        private final Set<Relation> collectedRelations = new HashSet<>();
+        @Override
+        public Void visitConstraint(Constraint constraint) {
+            for (Relation rel :  constraint.getConstrainedRelations()) {
+                if (collectedRelations.add(rel)) {
+                    rel.getDefinition().accept(this);
+                }
+            }
+            return null;
+        }
+    }
+
+    private Relation makePredefinedRelation(String name) {
+        /*
+            WARNING: The code has possibly unexpected behaviour:
+             - If a relation like `ext` is already defined (different from the standard definition),
+               constructing e.g. `rfe` will refer to this new definition.
+             - When constructing derived relations like fr, the intermediately defined relations like rf^-1
+               will get constructed even if they already exist in the model.
+             TODO: Clarify what the intended behaviour should be.
+         */
+        final Relation r = newRelation(name);
+        final Definition def = switch (name) {
             case PO -> new ProgramOrder(r, Filter.byTag(Tag.VISIBLE));
-            case LOC -> new SameAddress(r);
-            case ID -> new Identity(r, Filter.byTag(Tag.VISIBLE));
-            case INT -> new SameThread(r);
-            case EXT -> new DifferentThreads(r);
+            case LOC -> new SameLocation(r);
+            case ID -> new SetIdentity(r, Filter.byTag(Tag.VISIBLE));
+            case INT -> new Internal(r);
+            case EXT -> new External(r);
             case CO -> new Coherence(r);
             case RF -> new ReadFrom(r);
             case RMW -> new ReadModifyWrites(r);
-            case CASDEP -> new CompareAndSwapDependency(r);
-            case CRIT -> new CriticalSections(r);
+            case CASDEP -> new CASDependency(r);
+            case CRIT -> new LinuxCriticalSections(r);
             case IDD -> new DirectDataDependency(r);
             case ADDRDIRECT -> new DirectAddressDependency(r);
             case CTRLDIRECT -> new DirectControlDependency(r);
             case EMPTY -> new Empty(r);
             case RFINV -> {
                 //FIXME: We don't need a name for "rf^-1", this is a normal relation!
                 // This causes models that explicitly mention "rf^-1" to have two versions of the same relation!
-                yield new Inverse(r, getRelation(RF));
+                yield new Inverse(r, getOrCreatePredefinedRelation(RF));
             }
```

Comment:
Based on the comment ... do we even need to support this relation here?

Reply:
The reason we have this case is because of the definition of `FR`:
```
 case FR -> composition(r, getOrCreatePredefinedRelation(RFINV), getOrCreatePredefinedRelation(CO));
```
Now, I can change this definition to fetch `RF` and manually construct `RF^-1` from it.
This is fine as long as there is no other definition that wants to use `RF^-1` (because this would result in duplicated constructions).
I can do the same for `MM` and `MV` but not for, e.g., `IDDTRANS` because it is used in multiple places.